File: 프로그래머스/lv3/43238. 입국심사/입국심사.java
public class Solution {

    public static long solution(int n, int[] times){
        long start = 1;
        long end = 1_000_000_000_000_000_000L;

        while(end > start) {
            long mid = (start + end) / 2;

            if(isValid(n, mid, times)) {
                end = mid;
            }else {
                start = mid + 1;
            }
        }

        return start;
    }

    public static boolean isValid(int n, long t, int[] times) {
        long result = 0;
        for(int time : times) {
            result += t / time;
        }
        return result >= n;
    }

}

File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

long long solution(int n, vector<int> times) {
    long long answer = 0;
    
    long long max = ((long long)n) * *max_element(times.begin(), times.end());
    
    long long min = 0;
    long long mid;
    
    while(min < max)
    {
        mid = (max + min) / 2;
        
        long long total = 0;
        for(int i = 0; i < times.size(); ++i)
        {
            total += mid / times[i];
        }
        
        if (total >= n)
        {
            max = mid;
            answer = mid;
        }
        else
        {
            min = mid+1;
        }
    }
    
    
    return answer;
}